openraft/src/engine/log_id_list.rs

```diff
@@ -0,0 +1,105 @@
+use crate::raft_types::RaftLogId;
+use crate::LogId;
+use crate::NodeId;
+
+/// Efficient storage for log ids.
+///
+/// It stores only the ids of log that have a new leader_id. And the `last_log_id` at the end.
+/// I.e., the oldest log id belonging to every leader.
+///
+/// If it is not empty, the first one is `last_purged_log_id` and the last one is `last_log_id`.
+/// The last one may have the same leader id as the second last one.
+#[derive(Default, Debug, Clone)]
+#[derive(PartialEq, Eq)]
+pub struct LogIdList<NID: NodeId> {
+    key_log_ids: Vec<LogId<NID>>,
+}
+
+impl<NID: NodeId> LogIdList<NID> {
+    pub fn new(key_log_ids: impl IntoIterator<Item = LogId<NID>>) -> Self {
+        Self {
+            key_log_ids: key_log_ids.into_iter().collect(),
+        }
+    }
+
+    /// Extends a list of `log_id` that are proposed by a same leader.
+    ///
+    /// The log ids in the input has to be continuous.
+    pub(crate) fn extend_from_same_leader<'a, LID: RaftLogId<NID> + 'a>(&mut self, new_ids: &[LID]) {
+        if let Some(first) = new_ids.first() {
+            let first_id = first.get_log_id();
+            self.append(*first_id);
+
+            if let Some(last) = new_ids.last() {
+                let last_id = last.get_log_id();
+                assert_eq!(last_id.leader_id, first_id.leader_id);
+
+                if last_id != first_id {
+                    self.append(*last_id);
+                }
+            }
+        }
+    }
+
+    /// Append a new `log_id`.
+    ///
+    /// The log id to append does not have to be the next to the last one in `key_log_ids`.
+    /// In such case, it is meant to append a list of log ids.
+    ///
+    /// NOTE: The last two in `key_log_ids` may be with the same `leader_id`, because `last_log_id` always present in
+    /// `log_ids`.
+    pub(crate) fn append(&mut self, new_log_id: LogId<NID>) {
+        let l = self.key_log_ids.len();
+        if l == 0 {
+            self.key_log_ids.push(new_log_id);
+            return;
+        }
+
+        // l >= 1
+
```

Comment:
add assert of term too?

Reply:
i think `LogIdList` leak some check that: if append new id has a bigger term, say term `N+1`, `LogIdList` must check that:

* already has term `N` log saved.
* last log of term `N` index + 1 == first log of term `N+1`.

Reply:
Hm? It has already been checked that the new one is greater than the last one. Did you mean this?

Reply:
> i think `LogIdList` leak some check that: if append new id has a bigger term, say term `N+1`, `LogIdList` must check that:
> 
> * already has term `N` log saved.

The term does not need to be continuous. Just being incremental is enough.

> * last log of term `N` index + 1 == first log of term `N+1`.

This `append()` is optimized so that the index does not need to be continuous.
Appending log at index `i` implies appending all the log upto `i`.